internal/fs/caching_test.go

```diff
@@ -432,3 +436,313 @@ func (t *CachingWithImplicitDirsTest) SymlinksAreTypeCached() {
+	err = os.Mkdir(path.Join(bucket1MntDir, name), 0700)
+	AssertEq(nil, err)
+
+	// Remove the backing object in GCS.
+	err = uncachedBuckets[bucket1Name].DeleteObject(
+		ctx,
+		&gcs.DeleteObjectRequest{Name: name + "/"})
+
+	AssertEq(nil, err)
+
+	// Because we are caching, the directory should still appear to exist.
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectTrue(fi.IsDir())
+
+	// After the TTL elapses, we should see it disappear.
+	cacheClock.AdvanceTime(ttl + time.Millisecond)
+
+	_, err = os.Stat(path.Join(bucket1MntDir, name))
+	ExpectTrue(os.IsNotExist(err), "err: %v", err)
+
+	// AssertEq(nil, os.RemoveAll(path.Join(bucket1MntDir, name)))
+}
+
+func (t *MultiBucketMountCachingTest) ConflictingNames_RemoteModifier() {
+	const name = "foo"
+	var fi os.FileInfo
+	var err error
+	bucket1MntDir := getMultiMountBucketDir(bucket1Name)
+
+	// Create a directory via the file system.
+	err = os.Mkdir(path.Join(bucket1MntDir, name), 0700)
+	AssertEq(nil, err)
+
+	// Create a file with the same name via GCS.
+	_, err = storageutil.CreateObject(
+		ctx,
+		uncachedBuckets[bucket1Name],
+		name,
+		[]byte("taco"))
+
+	AssertEq(nil, err)
+
+	// Because the file system is caching types, it will fail to find the file
+	// when statting.
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectTrue(fi.IsDir())
+
+	_, err = os.Stat(path.Join(bucket1MntDir, name+inode.ConflictingFileNameSuffix))
+	ExpectTrue(os.IsNotExist(err), "err: %v", err)
+
+	// After the TTL elapses, we should see both.
+	cacheClock.AdvanceTime(ttl + time.Millisecond)
+
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectTrue(fi.IsDir())
+
+	fi, err = os.Stat(path.Join(bucket1MntDir, name+inode.ConflictingFileNameSuffix))
+	AssertEq(nil, err)
+	ExpectFalse(fi.IsDir())
+
+	AssertEq(nil, os.RemoveAll(path.Join(bucket1MntDir, name)))
+}
+
+func (t *MultiBucketMountCachingTest) TypeOfNameChanges_LocalModifier() {
+	const name = "test"
+	var fi os.FileInfo
+	var err error
+	bucket1MntDir := getMultiMountBucketDir(bucket1Name)
+
+	// Create a directory via the file system.
+	err = os.Mkdir(path.Join(bucket1MntDir, name), 0700)
+	AssertEq(nil, err)
+
+	// Delete it and recreate as a file.
+	err = os.Remove(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+
+	err = os.WriteFile(path.Join(bucket1MntDir, name), []byte("taco"), 0400)
+	AssertEq(nil, err)
+
+	// All caches should have been updated.
+	fi, err = os.Stat(path.Join(bucket1MntDir, name))
+	AssertEq(nil, err)
+	ExpectFalse(fi.IsDir())
+	ExpectEq(len("taco"), fi.Size())
+
+	AssertEq(nil, os.RemoveAll(path.Join(bucket1MntDir, name)))
+}
+
+func (t *MultiBucketMountCachingTest) TypeOfNameChanges_RemoteModifier() {
+	const name = "foo"
+	var fi os.FileInfo
+	var err error
+	bucket1MntDir := getMultiMountBucketDir(bucket1Name)
+	bucket1 := buckets[bucket1Name]
+
+	os.RemoveAll(path.Join(bucket1MntDir, name))
```

Comment:
nit: we can remove this if we implement tear down method ?

Reply:
A tear-down method would require us to maintain a list of files/directories/objects in struct/global-vars to be deleted, in this case. I think it is better to do deferred deletions within the test code itself, which is easier to both maintain and read. So, I've made changes taking that path. 

Let me know if you differ on that.

Reply:
@sethiay ^

Reply:
Same discussion as in https://github.com/GoogleCloudPlatform/gcsfuse/pull/1521#discussion_r1426214273 .